test/simulation.jl

```diff
@@ -0,0 +1,159 @@
+    statedefn = Ising2D(L; periodic)
+
+    sim = WangLandau.WangLandauSimulation(statedefn)
+    @test isa(sim, WangLandau.WangLandauSimulation)
+    @test isa(sim.logf_strategy, ReduceByFactor)
+    @test isa(sim.flat_strategy, FractionOfMean)           
+    @test isa(sim.catchup_strategy, NoCatchup)    
+
+    sim_check = WangLandau.WangLandauSimulation(statedefn; check_sweeps = 10)
+    @test sim_check.check_steps == 10 * WangLandau.system_size(statedefn)
+
+    sim_steps = WangLandau.WangLandauSimulation(statedefn; max_total_steps = 1)
+    @test_logs (:warn, r".*") WangLandau.WangLandauSimulation(statedefn; max_total_steps = 1)
+
+    sim_logf = WangLandau.WangLandauSimulation(statedefn; final_logf = 1e-3)
+    @test WangLandau.final_value(sim_logf.logf_strategy) == 1e-3
+
+    fs = FractionOfMean(0.75)
+    sim_flat = WangLandau.WangLandauSimulation(statedefn; flat_strategy = fs)
+    @test sim_flat.flat_strategy === fs
+
+    @test_throws ArgumentError WangLandau.WangLandauSimulation(statedefn; tasks_per_thread = -1)
+
+    io = IOBuffer()
+    show(io, sim)
+    out = String(take!(io))
+    @test occursin("WangLandauSimulation", out)
+    @test occursin("log(f)", out)
+    @test occursin("iterations", out)
+    @test occursin("total steps", out)
+    @test occursin("elapsed time", out)
+end    
+
+@testset "CommonSolve" begin
+    Random.seed!(1234)
+
+    L = 5
+    statedefn = Ising2D(L; periodic = false)
+    prob = WangLandauProblem(statedefn)
+
+    sim = init(prob; check_sweeps = 10, final_logf = 1e-3)
+
+    state, old_index = initialise_state(statedefn)
+    logdos = zeros(Float64, WangLandau.histogram_size(statedefn))
+    histogram = zeros(Int, WangLandau.histogram_size(statedefn))
+    logf = 0.1
+    catchup = NoCatchup()
+
+    new_index = WangLandau.wl_trial!(state, old_index, statedefn, logdos, histogram, logf, catchup)
+    @test 1 ≤ new_index ≤ length(histogram)
+    @test any(histogram .> 0)  
+    @test any(logdos .> 0)     
+
+    hist = zeros(Int, size(sim.samples))
+    task_samples = zeros(Int, max(1, sim.tasks_per_thread * Threads.nthreads()))
+    chunk_size = max(1, sim.check_steps ÷ length(task_samples))
+
+    flag = step!(sim, hist, task_samples, chunk_size)
+    @test sim.flat_checks ≥ 1
+    @test sim.total_steps ≥ sim.check_steps
+
+    sim_short = init(prob; check_sweeps = 5, final_logf = 1e-2, max_total_steps = 200)
+    sim_done = solve!(sim_short)
+    @test isa(sim_done, WangLandau.WangLandauSimulation)
+    @test sim_done.total_steps > 0
+    @test sim_done.elapsed_time ≥ 0.0
+end
+
+@testset "FixedFractionalCatchup" begin
+    Random.seed!(12345)
+
+    L = 5
+    periodic = false
+    statedefn = Ising2D(L; periodic)
+
+    ffc = FixedFractionalCatchup(0.25)
+    sim_stub = (; logdos = fill(1e-8, 4), catchup_strategy = ffc) 
+
+    sim_stub.logdos .= [1e-8, 2e-8, 3e-8, 4e-8][mod1.(1:length(sim_stub.logdos), 4)]
+
+    oldmin = ffc.minval
+    WangLandau.update!(ffc, sim_stub)
```

Comment:
We should have a test for `update!`

Reply:
I realised that there is a potential bug in `update!`. Have a look and please add a test for the case where `logdos` is all zeros